### PR TITLE
Add support for Call

### PR DESCRIPTION
**Pull request by iwahbe**
This PR adds support for the `Call` gRPC method at the top level `Provider` struct.

### CODE REVIEW

**Comment by t0yv0** [comment]
What's the test plan here?

**Comment by iwahbe** [comment]
> What's the test plan here?

I'm trying to determine the correct behavior before adding tests. We will definitely need tests before merging.

**Comment by iwahbe** [comment]
Tests have been added. I think this is ready to merge.

File: resourcex/traverse.go [hunk 1]
```diff
@@ -73,3 +73,47 @@ func Traverse(v resource.PropertyValue, p resource.PropertyPath, f func(resource
...
+
+// Walk traverses a property value along all paths, performing a depth first search.
+func Walk(v resource.PropertyValue, f func(resource.PropertyValue, WalkState)) {
+	walk(v, f, WalkState{})
+}
+
+type WalkState struct {
+	// We are entering the value. The children of the value have not yet been visited.
+	Entering bool
+	// The value is not transitively computed.
```
Comment (t0yv0) [inline resourcex/traverse.go:85]: This is a bit weird and does not mesh with my mental model:

```
type Value = Unknown
                    | Known ConcreteValue
```

Are there really transitively unknown values or this is just a degenerate encoding artifact that does not occur in practice?

File: provider.go [hunk 2]
```diff
@@ -718,8 +728,78 @@ func (p *provider) StreamInvoke(*rpc.InvokeRequest, rpc.ResourceProvider_StreamI
...
+		resourcex.Walk(v, func(v presource.PropertyValue, state resourcex.WalkState) {
+			if state.Entering || !v.IsOutput() {
+				return
+			}
+			for _, dep := range v.OutputValue().Dependencies {
+				urns = append(urns, string(dep))
+			}
+		})
+
+		returnDependencies[string(name)] = &rpc.CallResponse_ReturnDependencies{Urns: urns}
```
Comment (t0yv0) [inline provider.go:778]: Is this needed even when the provider has first-class outputs in Return, where this approximation can be un-abmiguously computed? Probably no harm in including but I thought maybe engine does this?

**Comment by t0yv0** [comment]
Nit: I would rename the PR to drop "at the provider level", looks like this introduces support for the Call gRPC method which is great.

**Comment by t0yv0** [comment]
One important question is instance lifetime. When a client issues a series of Call methods against a resource with the same URN, is there a possibility for the provider author to rely on the retention of in-memory instance state across the series of calls? 

```typescript
const c = new Counter()
c.increment()
c.increment()
export const n = c.curentValue /// 2
```

To achieve the above UX would it be possible for the implementation to be equally simple?

**Review by t0yv0 — APPROVED** [review]
Code looking good but I had a few non-blocking questions. I also would suggest considering an E2E test involving the real Pulumi CLI and a sample program, at least one of those  for the happy path.

**Comment by t0yv0** [comment]
Corner cases to think of (perhaps handled already in this codebase).

1. Does Call issue during preview. If so how are unknowns handled?

2. If there are schema-level or first-class secrets, does the implementation guarantee reasonable secret wrapping.

File: resourcex/traverse.go (hunk 1 above)
Comment (iwahbe) [reply resourcex/traverse.go:85]: I'm guessing it's a degenerate value, but lots of serde stuff requires valid types, even on computed fields so it might come up. This is mostly to be cautious.

File: provider.go [hunk 3]
```diff
@@ -718,8 +728,78 @@ func (p *provider) StreamInvoke(*rpc.InvokeRequest, rpc.ResourceProvider_StreamI
...
+	}
+
+	return &rpc.CallResponse{
+		Return:             _return,
+		ReturnDependencies: returnDependencies,
+		Failures:           checkFailureList(resp.Failures).rpc(),
+	}, nil
+}
+
+type CallRequest struct {
```
Comment (thomas11) [inline provider.go:793]: document public type?

File: provider.go [hunk 4]
```diff
@@ -718,8 +728,78 @@ func (p *provider) StreamInvoke(*rpc.InvokeRequest, rpc.ResourceProvider_StreamI
...
+		MonitorAddr:       req.GetMonitorEndpoint(),
+		Organization:      req.GetOrganization(),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to build pulumi.Context: %w", err)
+	}
+
+	args, err := p.getMap(req.GetArgs())
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert args into a property map")
```
Comment (thomas11) [inline provider.go:754]: [nit] include the original error to give an indication why conversion failed?

File: provider.go [hunk 5]
```diff
@@ -342,7 +343,11 @@ type Provider struct {
 	Read   func(context.Context, ReadRequest) (ReadResponse, error)
 	Update func(context.Context, UpdateRequest) (UpdateResponse, error)
 	Delete func(context.Context, DeleteRequest) error
-	// TODO Call
+
+	// Call allows methods to be attached to resources.
+	//
+	// Right now, Call is restricted to methods on component resources.
```
Comment (thomas11) [inline provider.go:349]: Is the last sentence still accurate? If so, maybe give readers a hint why that is.

File: provider.go (hunk 2 above)
Comment (iwahbe) [reply provider.go:778]: I'm not sure, but there is no documentation saying that it isn't necessary.

It seems like it is not necessary in Pulumi:

https://github.com/pulumi/pulumi/blob/cbfea508e7b2a4f1dd0985cefbeda3a6e891f7f9/pkg/resource/deploy/source_eval.go#L1140-L1142

I'm hesitant to assume that it is never necessary.

**Comment by iwahbe** [comment]
> Corner cases to think of (perhaps handled already in this codebase).
> 
> 1. Does Call issue during preview. If so how are unknowns handled?
> 2. If there are schema-level or first-class secrets, does the implementation guarantee reasonable secret wrapping.

1. I believe that it does. Call is supposed to result in a Pulumi program running (with `pulumi.Output`), so it unknowns should just flow through.
2. Not at all. This implementation just brings Call support into the base level of the library. Semantic guarentees like (if secret in the schema, then secret on the wire) are handled at a higher level (Call is not hooked up to `infer`).

File: provider.go (hunk 5 above)
Comment (iwahbe) [reply provider.go:349]: I linked the associated issues in pulumi/pulumi.

**Comment by iwahbe** [comment]
> One important question is instance lifetime. When a client issues a series of Call methods against a resource with the same URN, is there a possibility for the provider author to rely on the retention of in-memory instance state across the series of calls?
> 
> ```ts
> const c = new Counter()
> c.increment()
> c.increment()
> export const n = c.curentValue /// 2
> ```
> 
> To achieve the above UX would it be possible for the implementation to be equally simple?

I believe that this is possible now, though I would discourage relying on it. AFAIK the protocol does not make any guarantees about number of provider instances, or the order of calls. In theory, both `c.increment()`s could be called in parallel.

File: provider.go (hunk 2 above)
Comment (iwahbe) [reply provider.go:781]: I opened https://github.com/pulumi/pulumi/pull/17113 to pin down an answer here.
